java/feathub-udf/flink-udf/src/main/java/com/alibaba/feathub/flink/udf/processfunction/SlidingWindowKeyedProcessFunction.java

```diff
@@ -0,0 +1,190 @@
+ * for its window size. This will result in a lots of duplicates row in the state backend. For
+ * example, if we want to do aggregation under 1 hour and 2 hours, both sliding window will keep the
+ * rows in the last one hour, which are duplicated. With this process function, we only keep the
+ * rows for the maximum window size so that we can avoid duplicated rows in state backend.
+ *
+ * <p>The ProcessFunction assumes that: 1. rows of each key are ordered by the row time 2. row time
+ * attribute of rows with the same key are all distinct. The assumptions hold true after applying
+ * the tumbling window aggregation to the input with window size that is same as the step size of
+ * the {@link SlidingWindowKeyedProcessFunction} to be applied.
+ */
+public class SlidingWindowKeyedProcessFunction extends KeyedProcessFunction<Row, Row, Row> {
+
+    private final AggregationFieldsDescriptor aggregationFieldsDescriptor;
+    private final TypeSerializer<Row> rowTypeSerializer;
+    private final String rowTimeFieldName;
+    private final long stepSizeMs;
+    private MultiWindowSizeState state;
+
+    public SlidingWindowKeyedProcessFunction(
+            AggregationFieldsDescriptor aggregationFieldsDescriptor,
+            TypeSerializer<Row> rowTypeSerializer,
+            String rowTimeFieldName,
+            long stepSizeMs) {
+        this.aggregationFieldsDescriptor = aggregationFieldsDescriptor;
+        this.rowTypeSerializer = rowTypeSerializer;
+        this.rowTimeFieldName = rowTimeFieldName;
+        this.stepSizeMs = stepSizeMs;
+    }
+
+    @Override
+    public void open(Configuration parameters) {
+        state =
+                MultiWindowSizeState.buildMultiWindowSizeState(
+                        getRuntimeContext(), rowTypeSerializer);
+    }
+
+    @Override
+    public void processElement(
+            Row row, KeyedProcessFunction<Row, Row, Row>.Context ctx, Collector<Row> out)
+            throws Exception {
+        final long rowTime = ((Instant) row.getFieldAs(rowTimeFieldName)).toEpochMilli();
+        long triggerTime = rowTime;
+        while (triggerTime <= rowTime + aggregationFieldsDescriptor.getMaxWindowSizeMs()) {
+            ctx.timerService().registerEventTimeTimer(triggerTime);
+            triggerTime += stepSizeMs;
+        }
+        state.addRow(ctx.getCurrentKey(), rowTime, row);
+    }
+
+    @Override
+    public void onTimer(
+            long timestamp,
+            KeyedProcessFunction<Row, Row, Row>.OnTimerContext ctx,
+            Collector<Row> out)
+            throws Exception {
+        state.pruneRow(
+                ctx.getCurrentKey(), timestamp - aggregationFieldsDescriptor.getMaxWindowSizeMs());
+        aggregationFieldsDescriptor.getAggFieldDescriptors().forEach(d -> d.aggFunc.reset());
+
+        boolean hasRow = false;
+        for (long rowTime : state.orderedTimestampMap.get(ctx.getCurrentKey())) {
```

Comment:
Suppose window_size=3600 sec and step_size=1 sec, then we need to 3600 aggregation-operation values at each step.

Is it possible to use retraction so that we only need 1 aggregation operation at each step?

Which approach does Flink window uses? If it is the latter, then we might introduce regression here.

Reply:
I think the native Flink window operator uses a slightly different approach with the same time complexity. For each record, it is assigned and aggregated for every window it belongs to. [reference](https://stackoverflow.com/questions/64566540/apache-flink-incremental-window-computation). There is an [opensource project](https://tu-berlin-dima.github.io/scotty-window-processor/) that aims to solve the problem. 

I think we can certainly do the optimization in Feathub sliding window so I added a TODO.